### ISSUE TITLE
[Question] Model Interface: forward arguments

### ISSUE THREAD

**Issue by axelbr** [facebookresearch/mbrl-lib#101]
I am not sure how to use the forward function of the Model interface in my subclass as the interface only accepts a single input argument. If i want to have a control input, should i 1) concat the obs and action tensor, 2) use the TransitionBatch class or 3) introduce another function argument with default values (such as action=None)? What would be the preferred alternative for you?

Thanks!

**Comment by axelbr** [comment]
1. model. forward(x=torch.cat([obs, action])
2. model.forward(x=batch)
3. model.forward(x: TensorType, u: TensorType = None)

**Comment by luisenp** [comment]
Hi @axelbr. It looks like this is a place where the current documentation is a bit confusing. In fact, for the model `forward` method, you can actually use any number of input arguments. What needs to be defined to use the `ModelTrainer` is the `loss(model_input, optimizer, target=None)` method. This method is called by the default [update](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/models/model.py#L133) method, which is in turn [called](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/models/model_trainer.py#L119) by `ModelTrainer` using a `TransitionBatch` as model input (and note that it passes `target=None`). Here, there are two cases:

1. You want to have flexibility to use your model class with arbitrary tensors as inputs, w/o needed a transition batch every time. In this case, you can have another wrapper model that converts transition batch to tensors, creates the target, and then pass the correct tensor inputs and target to the model. This is the case for [GaussianMLP](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/models/gaussian_mlp.py#L294) (which only works with Tensors), and the wrapper [OneDTransitionRewardModel](https://github.com/facebookresearch/mbrl-lib/blob/master/mbrl/models/one_dim_tr_model.py#L172) (which converts `TransitionBatch` into `torch.cat([obs, action])` along with some other transformations, and creates the target from `batch.next_obs` and `batch.reward`.  

2. You are fine dealing with `TransitionBatch` directly. In this the model itself takes care of all the data manipulation and then can pass any tensors to `forward`. This is the case for the PlaNet [implementation](https://github.com/facebookresearch/mbrl-lib/blob/planet_impl/mbrl/models/planet.py#L261) we are currently working on.  Note that its forward method is defined as `forward(obs: torch.Tensor, action: torch.Tensor)`.

Does this make sense? Do let me know if this is still confusing, I'm happy to clarify further. BTW, if you are working with 1-D obs, and need obs and act as inputs, I suggest to rely on `OneDTransitionRewardModel` for batch manipulation, and have your model class just focus on the modeling part, dealing only with torch Tensors.

**Comment by axelbr** [comment]
Thanks for the extensive answer, it is clearer now. I just thought that implementations have to follow the interface more strictly. I will close this issue now. If further questions arise, I will post them here. 

Thanks for maintaining this codebase!
